bgpd/bgp_zebra.c

```diff
@@ -1378,6 +1379,15 @@ void bgp_zebra_announce(struct bgp_dest *dest, const struct prefix *p,
 		if (!nh_updated)
 			continue;
 
+		/* Allow recursion if it is a multipath group with both
+		 * eBGP and iBGP paths.
+		 */
```

Comment:
it seems potentially dangerous to allow recursion on both iBGP and eBGP routes, as eBGP routes can recurse onto iBGP routes ... I don't see how we can turn recursion on for one and not the other, however ...

Reply:
Recursive resolution is performed twice -- the first time when BGP is determining whether a route is valid. This is where you would expect eBGP routes not to resolve over iBGP routes. Then, when BGP sends the route to zebra here, it again requests (or not) that zebra restrict recursive resolution of nexthops that it puts in the FIB.

So at least in the obvious case, eBGP routes with nexthops resolved over iBGP routes don't even contribute to the nexthops being sent to zebra here. It's still more porous that it was before, no doubt. But as you say we turn recursion on for the whole route or not...